In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

In [2]:
train_df = pd.read_csv(Path('/home/dan/Documents/Repo/smu-dal-data-pt-09-2021-u-c/02-Homework/19-Supervised-Machine-Learning/Instructions/Resources/2019loans.csv'))
test_df = pd.read_csv(Path('/home/dan/Documents/Repo/smu-dal-data-pt-09-2021-u-c/02-Homework/19-Supervised-Machine-Learning/Instructions/Resources/2020Q1loans.csv'))

In [3]:
train_df.head()

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   

   annual_inc verification_status loan_status pymnt_plan  ...  pct_tl_nvr_dlq  \
0    223000.0        Not Verified    low_risk          n  ...           100.0   
1    123000.0     Source Verified    low_risk          n  ...            85.0   
2    197000.0     Source Verified    low_risk          n  ...            85.7   
3     45000.0        Not Verified    low_risk          n  ...           100.0   
4    133000.0     Source Verified    low_risk          n  ...           100.0   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              50.0                   0.0        0.0         577150.0   
1              33.3                   0.0        0.0         132750.0   
2              33.3                   0.0        0.0         628160.0   
3              16.7                   1.0        0.0          42006.0   
4              66.7                   0.0        0.0         283248.0   

   total_bal_ex_mort  total_bc_limit total_il_high_credit_limit  \
0           122018.0         32000.0                   170200.0   
1            27896.0         15900.0                    35398.0   
2           114043.0         22600.0                    90340.0   
3            20761.0         19900.0                    15406.0   
4           109056.0         79500.0                    58778.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  
2              N                     N  
3              N                     N  
4              N                     N  

[5 rows x 86 columns]

In [4]:
train_df['application_type']

0        Individual
1        Individual
2        Individual
3        Individual
4        Individual
            ...    
12175    Individual
12176    Individual
12177    Individual
12178     Joint App
12179    Individual
Name: application_type, Length: 12180, dtype: object

In [5]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12180 entries, 0 to 12179
Data columns (total 86 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  12180 non-null  int64  
 1   index                       12180 non-null  int64  
 2   loan_amnt                   12180 non-null  float64
 3   int_rate                    12180 non-null  float64
 4   installment                 12180 non-null  float64
 5   home_ownership              12180 non-null  object 
 6   annual_inc                  12180 non-null  float64
 7   verification_status         12180 non-null  object 
 8   loan_status                 12180 non-null  object 
 9   pymnt_plan                  12180 non-null  object 
 10  dti                         12180 non-null  float64
 11  delinq_2yrs                 12180 non-null  float64
 12  inq_last_6mths              12180 non-null  float64
 13  open_acc                    121

In [6]:
train_df.shape

(12180, 86)

In [7]:
data = pd.get_dummies(data = train_df, columns=['home_ownership', 'verification_status','pymnt_plan','initial_list_status','application_type','hardship_flag','debt_settlement_flag'])

In [8]:
train_df['loan_status']

0         low_risk
1         low_risk
2         low_risk
3         low_risk
4         low_risk
           ...    
12175    high_risk
12176    high_risk
12177    high_risk
12178    high_risk
12179    high_risk
Name: loan_status, Length: 12180, dtype: object

In [9]:
data['loan_status'] = data['loan_status'].astype('string')

In [10]:
loan = {'low_risk':1, 'high_risk':2}
data['loan_status'] = data['loan_status'].map(loan)

In [11]:
# Convert categorical data to numeric and separate tarclassifiert feature for testing data

X = data.drop(columns= ['loan_status'])
y = data['loan_status']

In [12]:
y

0        1
1        1
2        1
3        1
4        1
        ..
12175    2
12176    2
12177    2
12178    2
12179    2
Name: loan_status, Length: 12180, dtype: int64

In [13]:
# add missing dummy variables to testing set
test = pd.get_dummies(data = test_df, columns=['home_ownership', 'verification_status','pymnt_plan','initial_list_status','application_type','hardship_flag','debt_settlement_flag'])

In [32]:
test1 = test.drop(columns= ['loan_status'])
yset = test['loan_status']

In [21]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=42).fit(X, y)
print(f"Training Data Score: {classifier.score(X, y)}")

Training Data Score: 0.6485221674876848


/home/dan/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [22]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(max_depth=2, random_state=42)
RandomClassifier = clf.fit(X, y)
print(f"Training Data Score: {RandomClassifier.score(X, y)}")

Training Data Score: 0.6634646962233169


In [23]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_data = scaler.fit_transform(X)


In [24]:
# Train the Logistic Regression model on the scaled data and print the model score
from sklearn.linear_model import LogisticRegression
classifier1 = LogisticRegression(random_state=42).fit(X, y)
print(f"Training Data Score: {classifier1.score(scaled_data, y)}")

Training Data Score: 0.6386699507389163


/home/dan/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [28]:
# Train a Random Forest Classifier model on the scaled data and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(max_depth=8, random_state=42)
RandomClassifier1 = clf.fit(scaled_data, y)
print(f"Training Data Score: {RandomClassifier1.score(scaled_data, y)}")

Training Data Score: 0.8142036124794746


In [30]:
from sklearn.metrics import confusion_matrix

y_true = y
y_pred = classifier.predict(X)
confusion_matrix(y_true, y_pred)

array([[3657, 2433],
       [1848, 4242]])

In [31]:
tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
accuracy = (tp + tn) / (tp + fp + tn + fn) # (111 + 128) / (111 + 5 + 128 + 6)
print(f"Accuracy: {accuracy}")

Accuracy: 0.6485221674876848


In [37]:
#Applying the model on the test data
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=42).fit( test1, yset)
print(f"Test Data Score: {classifier.score(test1, yset)}")

Test Data Score: 0.8268821777966823


/home/dan/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [35]:
y_true = yset
y_pred = classifier.predict(test1)
confusion_matrix(y_true, y_pred)

array([[1760,  591],
       [ 223, 2128]])

#### Final Analysis 

The model final model scored 0.8268821777966823 on the test data.